# Secondary Notebook: Building A Corpus of Catalogue Entries
This secondary notebook serves to build a corpus of documents that will improve the gender component of my analysis. I aim at creating a corpus with all the entries of the 40 catalogues of our initial corpus. This larger corpus aims at improving the dictionary of women's name. In this case, the more documents we have to build our dictionary, the more chance we will have to correcftly identifiy the gender value of a document. I decided to do this in a secondary notebook in order to maintain the narrative flow of the main notebook.

To do so, I start from step [fill in] in the main Notebook. The catalogues imported have thus been cleaned once, and the same steps as Note book one will be completed in this notebook but for the word "Hatun", "Hâtun" and "bt." instead.


In [1]:
import os
import re

In [2]:
path = "/home/otzy/local_directory/Final Project/corpora/"
os.chdir(path)
filenames = os.listdir(path)
print(filenames)

['1613-1615 galata 37.txt', '1577-1578 galata 7.txt', '1573-1591 galata 15.txt', '1663-1663 galata 90.txt', '1675-1676 istanbul 18.txt', '1618-1618 istanbul 3.txt', '1524-1530 uskudar 5.txt', '1594-1595 rumeli 21.txt', '1670-1671 eyup 82.txt', '1612-1643 haskoy 5.txt', '1641-1644 galata 65.txt', '1647-1649 rumeli 80.txt', '1582-1583 uskudar 56.txt', '1633-1633 rumeli 56.txt', '1590-1591 uskudar 84.txt', '1637-1638 eyup 37.txt', '1579-1580 uskudar 51.txt', '1655-1655 eyup 61.txt', '1546-1549 uskudar 2.txt', '1585-1587 eyup 3.txt', '1606-1607 galata 32.txt', '1691-1691 bab 54.txt', '1726-1738 istanbul 24.txt', '1679-1680 eyup 90.txt', '1615-1620 galata 46.txt', '1644-1644 eyup 49.txt', '1546-1549 uskudar 14.txt', '1563-1563 balat 2.txt', '1619-1620 eyup 19.txt', '1663-1664 istanbul 12.txt', '1685-1686 bab 46.txt', '1549-1556 uskudar 17 .txt', '1575-1576 Galata 5.txt', '1666-1667 bab 3.txt', '1534-1536 uskudar 9 .txt', '1674-1679 haskoy 10.txt', '1661-1662 eyup 74.txt', '1562-1563 uskudar

So for this time, our corpus is all of the catalogues entries. So thats about it, we can now loop through our files to uncover all possible woman names.

In [3]:
paragraphs = [] # track paragraphs
for filename in filenames:
    with open(path+filename, 'r') as f:
        string = f.read()
    pos = 0 # initial position marker (assuming our keyword isn't at the very start)
    while(pos>-1): # loop while we have a match
        pos = string.find("hâtun", pos+1) # find next match
        leftNewLine = string.rfind("\n\n", 0, pos) # find previous newline
        rightNewLine = string.find("\n\n", pos) # find next newline
        paragraph = string[leftNewLine+2:rightNewLine] # isolate paragraph
        if paragraph: # if we have a non-empty string
            paragraphs.append(paragraph) # add it to the paragraphs
    while(pos>-1): # loop while we have a match
        pos = string.find("hatun", pos+1) # find next match
        leftNewLine = string.rfind("\n\n", 0, pos) # find previous newline
        rightNewLine = string.find("\n\n", pos) # find next newline
        paragraph = string[leftNewLine+2:rightNewLine] # isolate paragraph
        if paragraph: # if we have a non-empty string
            paragraphs.append(paragraph) # add it to the paragraphs
    while(pos>-1): # loop while we have a match
        pos = string.find("Hâtun", pos+1) # find next match
        leftNewLine = string.rfind("\n\n", 0, pos) # find previous newline
        rightNewLine = string.find("\n\n", pos) # find next newline
        paragraph = string[leftNewLine+2:rightNewLine] # isolate paragraph
        if paragraph: # if we have a non-empty string
            paragraphs.append(paragraph) # add it to the paragraphs
    while(pos>-1): # loop while we have a match
        pos = string.find("Hatun", pos+1) # find next match
        leftNewLine = string.rfind("\n\n", 0, pos) # find previous newline
        rightNewLine = string.find("\n\n", pos) # find next newline
        paragraph = string[leftNewLine+2:rightNewLine] # isolate paragraph
        if paragraph: # if we have a non-empty string
            paragraphs.append(paragraph) # add it to the paragraphs
os.mkdir(path+'kadinlara ilgili kaynaklar')
os.chdir(path+'kadinlara ilgili kaynaklar')
for i, paragraph in enumerate(paragraphs):
    filename = '{0:03d}'.format(i)+'.txt'
    with open(filename, 'w') as f:
        f.write(paragraph)

In [ ]:
print(len(paragraphs))

4665


Great, we have 4665 entries that contain the word Hatun (Lady), which most often comes with a women's name.

Let's now extract these names. To do so I propose to craft our own tokenizer built in order to produce tokens with the words around Hatun. Thanks to the transliterators from Arabic to the Latin script, the proper names are identified with a capital letter! It's easier to tokenize the right one and to build the dictionary.

In [ ]:
from nltk.tokenize import RegexpTokenizer
import nltk
women_nametokens = RegexpTokenizer(r"(?:[A-Z]\w+\s[A-Z]\w+\sbt.)|(?:[A-Z]\w+\s(?:Hatun|Hâtun))") 
#RegexpTokenizer isn't something we have studied in class!
#(?:[regex]) is a non-capturing group as required by the program. See link here:
#http://www.nltk.org/api/nltk.tokenize.html?highlight=tokenizer#module-nltk.tokenize.regexp
from nltk.corpus import PlaintextCorpusReader
path = "/home/otzy/local_directory/Final Project/corpora/kadinlara ilgili kaynaklar/"
kadinCorpus = PlaintextCorpusReader(path, '.*txt')
string = ""
for fileid in kadinCorpus.fileids():
    text = kadinCorpus.raw(fileid)# read each text and convert to lower case
    tokens = women_nametokens.tokenize(text) # tokenize
    tokens = " ".join(tokens)
    string += tokens
#cleans the irregularities of string
#Let's clean the list so that only the names remain, that 
#is, let us remove "bt." and "Hatun"(both being marks of respect for ladies). 
#Afterwards, we'll be able to build a dictionary of only women names.
string = re.sub("Hatun|Hâtun|bt.", "", string) 
print(type(string))

Let's turn this string into a dictionary in order to attribute gender values to particular documents.

In [ ]:
#transforms string back into list and dictionary
women_names = nltk.word_tokenize(string)
#create dictionary of names
dict_women_names= {} # build a compilation of all of women's names
#if not already in dictionary, add women_names as dict_women_names key
for token in women_names:
    if token not in dict_women_names:
        dict_women_names[token] = 1 #gives value 1
print(dict_women_names)
print(len(dict_women_names))

For any words that would create confusion(i.e. gender neutral, polysemic proper nouns, etc.), let's read the list, identify a few problematic variable, and use the dict.pop(arg1) to remove the undesirable arguments.

In [ ]:
entries = ("Ermeni", "Emir", "Ayni", "Selçuk", "Envar", "Ali", "Edirneli", "Musâhibe", "Bahri", "Bâdıseher", "Mezbûre", "Usul", "Lütfi", "Melek", "Tomay", "Hammâmî", "Paşa", "Güllü", "Merhûm", "Belgradlı", "Hayati", "Ziyâde", "Ziyade", "Hoca", "Gülüm", "Abdülkerim", "Nâme", "Akpula", "Saadet", "Toklu", "Can", "Dilsiz", "Papa", "Reis", "Bey", "Murad", "Ağa", "Ramazan", "Hasan", "Beşe", "Süleyman", "Cihan", "Evin", "Boşanan", "Timurhan", "Tâcî", "Sultan", "Gevher")
#With some googling, it's easy to find inspiration to delete the entries enumerated above!
#the solution I found proposed
#thanks to https://stackoverflow.com/questions/8995611/removing-multiple-keys-from-a-dictionary-safely/30351294
#defining a function isn't something we have done much in class!
def entries_to_remove(entries, dict_women_names): 
    for key in entries:
        if key in dict_women_names:
            del dict_women_names[key]
entries_to_remove(entries, dict_women_names)
print(len(dict_women_names))

In [ ]:
print(dict_women_names)

There we go, our dictioanry is now ready to be added to the gender dictionary of the main notebook.